apps/api/src/app/environments-v1/usecases/outputRendererUseCases.ts

```diff
@@ -0,0 +1,144 @@
+  RedirectTargetEnum,
+  SmsRenderOutput,
+  TipTapNode,
+} from '@novu/shared';
+import { z } from 'zod';
+import { ExpendEmailEditorSchemaUseCase } from '../../environments/render/email-schema-extender';
+
+class RenderCommand {
+  controlValues: Record<string, unknown>;
+}
+
+export class EmailOutputRendererUseCase {
+  async execute(renderCommand: RenderCommand): Promise<EmailRenderOutput> {
+    const parse = EmailStepControlSchema.parse(renderCommand.controlValues);
+    const schema = parse.emailEditor as TipTapNode;
+    const expandedSchema = new ExpendEmailEditorSchemaUseCase().execute({ schema });
+    const html = await render(expandedSchema);
+
+    return { subject: parse.subject, body: html };
+  }
+}
+
+// Concrete Renderer for Chat Preview
+export class ChatOutputRendererUseCase {
+  execute(renderCommand: RenderCommand): ChatRenderOutput {
+    const body = (renderCommand.controlValues.body as string) || 'Default chat message';
+
+    return { body };
+  }
+}
+
+// Concrete Renderer for SMS Preview
+export class SmsOutputRendererUseCase {
+  execute(renderCommand: RenderCommand): SmsRenderOutput {
+    const body = (renderCommand.controlValues.body as string) || 'Default SMS message';
+
+    return { body };
+  }
+}
+
+// Concrete Renderer for Push Notification Preview
+export class PushOutputRendererUseCase {
+  execute(renderCommand: RenderCommand): PushRenderOutput {
+    const subject = (renderCommand.controlValues.subject as string) || 'Default Push Notification Subject';
+    const body = (renderCommand.controlValues.body as string) || 'Default Push Notification Body';
+
+    return { subject, body };
+  }
+}
+
+// Concrete Renderer for In-App Message Preview
+export class InAppOutputRendererUseCase {
+  execute(renderCommand: RenderCommand): InAppRenderOutput {
+    console.log('InAppOutputRendererUseCase.execute', JSON.stringify(renderCommand, null, 2));
+    const inApp = InAppRenderOutputSchema.parse(renderCommand.controlValues);
+
+    return {
+      subject: inApp.subject as string,
+      body: inApp.body as string,
+      avatar: inApp.avatar as string,
+      primaryAction: {
+        label: inApp.primaryAction.label as string,
+        redirect: {
+          url: inApp.primaryAction.redirect.url as string,
+          target: inApp.primaryAction.redirect.target as RedirectTargetEnum,
+        },
+      },
+      secondaryAction: (inApp.secondaryAction?.label as string)
+        ? {
+            label: inApp.secondaryAction?.label as string,
+            redirect: {
+              url: inApp.secondaryAction?.redirect.url as string,
+              target: inApp.secondaryAction?.redirect.target as RedirectTargetEnum,
+            },
+          }
+        : undefined,
+      redirect: {
+        url: inApp.redirect.url as string,
+        target: inApp.redirect.target as RedirectTargetEnum,
+      },
+      data: inApp.data as Record<string, unknown>,
+    };
+  }
+}
+const emailContentSchema = z
```

Comment:
This schema seems to be the same as the step-controls endpoint needs to generate the default schemas. I am a bit surprised we define them for the first time here. Am I missing something?

Reply:
it's a diffrent story, this is a typed schema, I'm not sure there is a point to make the endpoint ( which we need to remove BTW ) return a schema for tiptap shcema, it's a single element producing a single control value.